allow multiple commands to run on build

nikolaistraessle opened the pull request:
Currently, it is not possible to run multiple commands with the flag `--build`.

With this change, it is possible to run multiple commands defined in the `--build` flag.

The commands must be separated by a double ampersand (`&&`).

For example, `CompileDaemon --build="go mod vendor && go build"`.

githubnemo commented:
Hi, thanks for the contribution!

I think it is bad practice to suggest the presence of a well-known feature when it is, in fact, not that feature - in this case command chaining as implemented by *sh.

I think I would rather have multiple `--build` arguments for multiple commands than add complexity in the form of string parsing (and escaping of special character sequences like `&&`).

The invocation would look like this:

```
CompileDaemon --build "go mod vendor" --build "go build" .
```

This would also not add the ability to do something like `--build "go mod vendor || (echo "no update" && go build)"` but it also does not suggest that it is capable of doing that. For more complex cases I would always recommend to use shell scripts and execute them instead.

What do you think?

githubnemo commented on daemon.go:
```diff
@@ -151,29 +151,37 @@ func failColor(format string, args ...interface{}) string {
+		c = strings.TrimSpace(c)
+		args := strings.Split(c, " ")
+		if len(args) == 0 {
+			// If the user has specified and empty then we are done.
```
This comment does not make sense

nikolaistraessle replied:
Yes, that comment make no sense, but this was from another contributor.

githubnemo commented on daemon.go:
```diff
@@ -151,29 +151,37 @@ func failColor(format string, args ...interface{}) string {
+			log.Println(failColor("Error while building:\n"), failColor(string(output)))
+			if success {
+				success = false
+			}
```
This should just be an early return - when one build command fails, execution should be stopped.

githubnemo commented on daemon.go:
```diff
@@ -151,29 +151,37 @@ func failColor(format string, args ...interface{}) string {
-	}
+	commands := strings.Split(*flagBuild, "&&")
+	success := true
+	for _, c := range commands {
```
I would opt to refactor the inner part of this loop into its own function for clarity.

nikolaistraessle commented on daemon.go:
```diff
@@ -151,29 +151,37 @@ func failColor(format string, args ...interface{}) string {
-		return true
-	}
+	commands := strings.Split(*flagBuild, "&&")
+	success := true
```
Makes sense.

nikolaistraessle commented:
@githubnemo many thanks for your review and suggestions for improvement.

I've changed to code to allow using the `--build` flag multiple times instead of using the shell separator `&&`.
Also, I extracted the functionality to run a build command into a separate method.

githubnemo commented on daemon.go:
```diff
@@ -375,6 +398,7 @@ func main() {
 	flag.Var(&flagExcludedDirs, "exclude-dir", " Don't watch directories matching this name, can be set more than once")
 	flag.Var(&flagExcludedFiles, "exclude", " Don't watch files matching this name, can be set more than once")
 	flag.Var(&flagIncludedFiles, "include", " Watch files matching this name, can be set more than once")
+	flag.Var(&flagBuildCommandList, "build", "Command to rebuild after changes, can be set more than once")
```
```suggestion
	flag.Var(&flagBuildCommandList, "build", "Command to rebuild after changes, can be set more than once. Defaults to 'go build'.")
```

githubnemo commented on daemon.go:
```diff
@@ -167,13 +192,11 @@ func build() bool {
 	}
-
-	return err == nil
+	return true
```
I think I would have returned the error and did the logging one level higher up for better separation of concerns.

nikolaistraessle replied:
I've added a success log message if a command was success. The executed command will also be logged in both case (error/success).
Now you can see in the output, which command has failed or was success.

githubnemo reviewed (commented):
Looks good, thanks! Some more comments.

nikolaistraessle commented:
@githubnemo I've added some more log events.

githubnemo commented on daemon.go:
```diff
@@ -149,12 +162,26 @@ func failColor(format string, args ...interface{}) string {
+			log.Println(failColor("Command failed:\n"), failColor(c))
+			return false
+		}
+		log.Println(okColor("Command success:\n"), okColor(c))
```
```suggestion
```
I would opt to remove this - this info is only relevant if something fails.

githubnemo commented on daemon.go:
```diff
@@ -149,12 +162,26 @@ func failColor(format string, args ...interface{}) string {
+	for _, c := range flagBuildCommandList.commands {
+		err := runBuildCommand(c)
+		if err != nil {
+			log.Println(failColor("Command failed:\n"), failColor(c))
```
```suggestion
			log.Println(failColor("Command failed: "), failColor(c))
```

githubnemo reviewed (commented):
Two minor things then this is ready to be merged :)

nikolaistraessle commented:
@githubnemo thanks for your change suggestions.

githubnemo commented:
Great addition, thanks for your work! :)
